File: validator/server_common/machine_locator.go
package server_common

import (
	"errors"
	"os"
	"path/filepath"
	"runtime"
	"strings"

	"github.com/ethereum/go-ethereum/common"
	"github.com/ethereum/go-ethereum/log"
)

type MachineLocator struct {
	rootPath    string
	latest      common.Hash
	moduleRoots []common.Hash
}

var ErrMachineNotFound = errors.New("machine not found")

func NewMachineLocator(rootPath string) (*MachineLocator, error) {
	dirs := []string{rootPath}
	if rootPath == "" {
		// Check the project dir: <project>/arbnode/node.go => ../../target/machines
		_, thisFile, _, ok := runtime.Caller(0)
		if !ok {
			panic("failed to find root path")
		}
		projectDir := filepath.Dir(filepath.Dir(filepath.Dir(thisFile)))
		projectPath := filepath.Join(filepath.Join(projectDir, "target"), "machines")
		dirs = append(dirs, projectPath)

		// Check the working directory: ./machines and ./target/machines
		workDir, err := os.Getwd()
		if err != nil {
			return nil, err
		}
		workPath1 := filepath.Join(workDir, "machines")
		workPath2 := filepath.Join(filepath.Join(workDir, "target"), "machines")
		dirs = append(dirs, workPath1)
		dirs = append(dirs, workPath2)

		// Check above the executable: <binary> => ../../machines
		execfile, err := os.Executable()
		if err != nil {
			return nil, err
		}
		execPath := filepath.Join(filepath.Dir(filepath.Dir(execfile)), "machines")
		dirs = append(dirs, execPath)
	}

	var (
		moduleRoots      = make(map[common.Hash]bool)
		latestModuleRoot common.Hash
	)

	for _, dir := range dirs {
		fInfo, err := os.Stat(dir)
		if err != nil {
			log.Warn("Getting file info", "error", err)
			continue
		}
		if !fInfo.IsDir() {
			// Skip files that are not directories.
			continue
		}
		files, err := os.ReadDir(dir)
		if err != nil {
			log.Warn("Reading directory", "dir", dir, "error", err)
		}
		for _, file := range files {
			mrFile := filepath.Join(dir, file.Name(), "module-root.txt")
			if _, err := os.Stat(mrFile); err != nil {
				// Skip if module-roots file does not exist.
				continue
			}
			mrContent, err := os.ReadFile(mrFile)
			if err != nil {
				log.Warn("Reading module roots file", "file path", mrFile, "error", err)
				continue
			}
			moduleRoot := common.HexToHash(strings.TrimSpace(string(mrContent)))
			if file.Name() != "latest" && file.Name() != moduleRoot.Hex() {
				continue
			}
			moduleRoots[moduleRoot] = true
			if file.Name() == "latest" {
				latestModuleRoot = moduleRoot
			}
			rootPath = dir
		}
		if rootPath != "" {
			break
		}
	}
	var roots []common.Hash
	for k := range moduleRoots {
		roots = append(roots, k)
	}
	return &MachineLocator{
		rootPath:    rootPath,
		latest:      latestModuleRoot,
		moduleRoots: roots,
	}, nil
}

func (l MachineLocator) GetMachinePath(moduleRoot common.Hash) string {
	if moduleRoot == (common.Hash{}) || moduleRoot == l.latest {
		return filepath.Join(l.rootPath, "latest")
	} else {
		return filepath.Join(l.rootPath, moduleRoot.String())
	}
}

func (l MachineLocator) LatestWasmModuleRoot() common.Hash {
	return l.latest
}

func (l MachineLocator) RootPath() string {
	return l.rootPath
}

func (l MachineLocator) ModuleRoots() []common.Hash {
	return l.moduleRoots
}

File: crates/validator/src/server_jit/machine_locator.rs
use anyhow::Result;
use arbutil::Bytes32;
use std::collections::HashSet;
use std::env;
use std::fs;
use std::path::{Path, PathBuf};
use std::str::FromStr;
use tracing::debug;
use tracing::warn;

use crate::config::ModuleRootConfig;

#[derive(Debug, Clone)]
pub struct MachineLocator {
    root_path: PathBuf,
    latest: Bytes32,
    module_roots: Vec<Bytes32>,
}

impl MachineLocator {
    pub fn new(module_root_config: &ModuleRootConfig) -> Result<Self> {
        let mut dirs = Vec::new();
        let mut module_roots_set = HashSet::new();
        let mut latest_module_root = Bytes32::default();
        let mut final_root_path = PathBuf::new();

        // Use CARGO_MANIFEST_DIR to find the crate root.
        let manifest_dir = Path::new(env!("CARGO_MANIFEST_DIR"));

        if let Some(module_root) = module_root_config.module_root {
            module_roots_set.insert(module_root);
            latest_module_root = module_root;

            final_root_path =
                if let Some(grandparent) = manifest_dir.parent().and_then(|p| p.parent()) {
                    grandparent.join("target").join("machines")
                } else {
                    final_root_path
                };
        } else {
            if let Some(rp) = module_root_config
                .module_root_path
                .as_ref()
                .filter(|s| s.exists())
            {
                dirs.push(PathBuf::from(rp));
            } else {
                // Try to find the workspace root by looking for "target" in common locations
                // <crate_root>/../../target/machines
                if let Some(grandparent) = manifest_dir.parent().and_then(|p| p.parent()) {
                    dirs.push(grandparent.join("target").join("machines"));
                }
                // <crate_root>/target/machines
                dirs.push(manifest_dir.join("target").join("machines"));

                // Check working directory
                if let Ok(work_dir) = env::current_dir() {
                    dirs.push(work_dir.join("machines"));
                    dirs.push(work_dir.join("target").join("machines"));
                }

                // Check relative to executable
                if let Ok(exec_path) = env::current_exe() {
                    if let Some(grandparent_of_exec) = exec_path.parent().and_then(|p| p.parent()) {
                        dirs.push(grandparent_of_exec.join("machines"));
                    }
                }
            }
        }

        for dir in dirs {
            if !dir.exists() || !dir.is_dir() {
                debug!("{dir:?} does not exist!!!");
                continue;
            }

            let entries = match fs::read_dir(&dir) {
                Ok(e) => e,
                Err(e) => {
                    warn!("Reading directory {:?} error: {}", dir, e);
                    continue;
                }
            };

            for entry in entries.flatten() {
                let mr_file = entry.path().join("module-root.txt");

                if !mr_file.exists() {
                    continue;
                }

                let mr_content = match fs::read_to_string(&mr_file) {
                    Ok(c) => c,
                    Err(e) => {
                        warn!("Reading module roots file {:?} error: {}", mr_file, e);
                        continue;
                    }
                };

                let module_root = match Bytes32::from_str(mr_content.trim()) {
                    Ok(h) => h,
                    Err(_) => {
                        warn!("Error converting module root file {mr_file:?} into hash");
                        continue;
                    }
                };

                let dir_name = entry.file_name().to_string_lossy().to_string();

                // IMPORTANT:
                // Go's moduleRoot.Hex() returns "0x" + hex.
                // Rust Bytes32 Display impl returns raw hex.
                // We must format it manually to match Go's directory naming convention.
                let module_root_hex = format!("0x{}", module_root);

                if dir_name != "latest" && dir_name != module_root_hex {
                    continue;
                }

                module_roots_set.insert(module_root);

                if dir_name == "latest" {
                    latest_module_root = module_root;
                }

                final_root_path = dir.canonicalize().unwrap_or(dir.clone());
            }

            if !final_root_path.as_os_str().is_empty() {
                break;
            }
        }

        let module_roots: Vec<Bytes32> = module_roots_set.into_iter().collect();

        Ok(MachineLocator {
            root_path: final_root_path,
            latest: latest_module_root,
            module_roots,
        })
    }

    pub fn get_machine_path(&self, module_root: Bytes32) -> PathBuf {
        if module_root == Bytes32::default() || module_root == self.latest {
            self.root_path.join("latest")
        } else {
            self.root_path.join(format!("0x{}", module_root))
        }
    }

    pub fn latest_wasm_module_root(&self) -> Bytes32 {
        self.latest
    }

    pub fn module_roots(&self) -> &[Bytes32] {
        &self.module_roots
    }
}

#[cfg(test)]
mod tests {
    use std::str::FromStr;

    use arbutil::Bytes32;

    use crate::{config::ModuleRootConfig, server_jit::machine_locator::MachineLocator};

    #[test]
    fn test_new_machine_locator_with_path() {
        let locator_path = "testdata".to_owned();
        let config = ModuleRootConfig {
            module_root: None,
            module_root_path: Some(locator_path.into()),
        };
        let machine_locator = MachineLocator::new(&config).unwrap();

        let expected_latest =
            Bytes32::from_str("0xf4389b835497a910d7ba3ebfb77aa93da985634f3c052de1290360635be40c4a")
                .unwrap();
        assert_eq!(expected_latest, machine_locator.latest);

        let expected_module_roots = [
            Bytes32::from_str("0x68e4fe5023f792d4ef584796c84d710303a5e12ea02d6e37e2b5e9c4332507c4")
                .unwrap(),
            Bytes32::from_str("0x8b104a2e80ac6165dc58b9048de12f301d70b02a0ab51396c22b4b4b802a16a4")
                .unwrap(),
            Bytes32::from_str("0xf4389b835497a910d7ba3ebfb77aa93da985634f3c052de1290360635be40c4a")
                .unwrap(),
        ];

        let mut module_roots = machine_locator.module_roots().to_vec();
        module_roots.sort();
        let module_roots: [Bytes32; 3] = module_roots.try_into().unwrap();

        assert_eq!(expected_module_roots, module_roots);
    }

    #[test]
    fn test_new_machine_locator_with_module_root() {
        let config = ModuleRootConfig {
            module_root: Some(
                Bytes32::from_str(
                    "0x68e4fe5023f792d4ef584796c84d710303a5e12ea02d6e37e2b5e9c4332507c4",
                )
                .unwrap(),
            ),
            module_root_path: None,
        };
        let machine_locator = MachineLocator::new(&config).unwrap();

        let expected_latest =
            Bytes32::from_str("0x68e4fe5023f792d4ef584796c84d710303a5e12ea02d6e37e2b5e9c4332507c4")
                .unwrap();
        assert_eq!(expected_latest, machine_locator.latest);

        let expected_module_roots = [Bytes32::from_str(
            "0x68e4fe5023f792d4ef584796c84d710303a5e12ea02d6e37e2b5e9c4332507c4",
        )
        .unwrap()];

        let mut module_roots = machine_locator.module_roots().to_vec();
        module_roots.sort();
        let module_roots: [Bytes32; 1] = module_roots.try_into().unwrap();

        assert_eq!(expected_module_roots, module_roots);
    }
}
